contracts/distribution/dao-rewards-distributor/src/contract.rs

```diff
@@ -0,0 +1,683 @@
+            } else {
+                Ok(0)
+            }
+        }
+        (Expiration::Never {}, Expiration::Never {}) => Ok(0),
+        _ => Err(StdError::generic_err(format!(
+            "incompatible expirations: got a {:?}, b {:?}",
+            a, b
+        ))),
+    }
+}
+
+fn get_rewards_per_token(
+    reward_config: &DenomRewardConfig,
+    env: &Env,
+    vp_contract: &Addr,
+    deps: Deps,
+) -> StdResult<Uint256> {
+    let crpt = CUMULATIVE_REWARDS_PER_TOKEN.load(deps.storage, reward_config.to_str_denom())?;
+
+    // query the current total voting power from the voting power contract
+    let total_power = get_total_voting_power(deps, env, vp_contract)?;
+
+    let last_time_reward_applicable =
+        reward_config.get_latest_reward_distribution_expiration_date(&env.block);
+
+    let expiration_diff = Uint128::from(get_expiration_diff(
+        last_time_reward_applicable,
+        reward_config.last_update,
+    )?);
+
+    if total_power.is_zero() {
+        Ok(crpt)
+    } else {
+        let reward_rate_time_units =
+            get_duration_units(&reward_config.reward_emission_config.reward_rate_time);
+
+        let complete_distribution_periods =
+            expiration_diff.checked_div(Uint128::from(reward_rate_time_units))?;
+
+        let numerator = reward_config
+            .reward_emission_config
+            .reward_rate_emission
+            .full_mul(complete_distribution_periods)
+            .checked_mul(scale_factor())?;
+        let denominator = Uint256::from(total_power);
+        Ok(crpt + numerator.checked_div(denominator)?)
+    }
+}
+
+fn get_accrued_rewards_since_last_user_action(
+    deps: Deps,
+    env: &Env,
+    addr: &Addr,
+    reward_per_token: Uint256,
+    vp_contract: &Addr,
+    denoms: Vec<String>,
+) -> StdResult<HashMap<String, Uint128>> {
+    // Get the user's voting power at the current height.
+    let voting_power = Uint256::from(get_voting_power(deps, env, vp_contract, addr)?);
+
+    let mut entitled_rewards: HashMap<String, Uint128> = HashMap::new();
+
+    let user_reward_config = USER_REWARD_CONFIGS
+        .load(deps.storage, addr.clone())
+        .unwrap_or_default();
+
+    for denom in denoms.iter() {
+        let user_last_reward_per_token = user_reward_config
+            .user_reward_per_token
+            .get(denom)
+            .cloned()
+            .unwrap_or_default();
+
+        // Calculate the difference between the current reward per token value and the user's latest reward per token value
+        let reward_factor = reward_per_token.checked_sub(user_last_reward_per_token)?;
+
+        // Calculate the amount of rewards earned.
+        // voting_power * reward_factor / scale_factor
+        let earned_rewards_amount: Uint128 = voting_power
+            .checked_mul(reward_factor)?
+            .checked_div(scale_factor())?
+            .try_into()?;
+
+        entitled_rewards.insert(denom.to_string(), earned_rewards_amount);
```

Comment:
i must be misunderstanding how this `get_accrued_rewards_since_last_user_action` works, but it seems like this is using the current voting power to determine all of the previously unclaimed rewards.

what i'm imagining:
- user has 10 voting power
- reward distribution starts
- 1 week goes by
- user stakes to get 20 voting power
- hook is triggered to update rewards for the user
- all of the unclaimed tokens (`reward_factor`) from that period get multiplied by the new voting power (20) and saved as pending

what am i missing?

Reply:
great question, i believe the way it works is:
1. user has 10 voting power and we start the rewards distribution
2. 1 week goes by, user accrued X rewards, but that is not logged anywhere yet
3. user doubles their stake to 20 voting power, `execute_stake` hook triggers `update_rewards`
4. `update_rewards` then:
    - gets `get_total_earned_puvp`. because it's the first time it's being called,
    it calculates the total rewards earned per unit voting power for duration from the last update
    until this block, which is that of 1 week.
    - stores that `total_earned_puvp` in the denom state. this is a global config that affects other users:
        - given that user now doubled their stake, other users will get relatively less rewards
        - updating the global config as a result of this action ensures that the next state change hook
        (from any user) will have the latest "snapshot" of the stakers' voting powers.
        i'll try to go deep on this point in the docs.
    - finds the user's accrued rewards since the last user action.
    because it's the first time that user is staking since rewards started,
    the user's denom_rewards_puvp for this denom is 0. users total rewards are
    therefore X - 0, because 0 is the last known users denom_rewards_puvp.
5. we update the user's reward state:
  - previous pending rewards: 0
  - earned rewards: X
  - new pending rewards: 0 + X = X

I have added a [unit test](https://github.com/DA0-DA0/dao-contracts/blob/0e28d29faa233d78239fb28e211d6bab4c63bcb7/contracts/distribution/dao-rewards-distributor/src/testing/tests.rs#L969) that goes through this in practice:

```rust

#[test]
fn test_native_dao_rewards_entry_edge_case() {
    let mut suite = SuiteBuilder::base(super::suite::DaoType::Native).build();

    // we start with the following staking power split:
    // [ADDR1: 100, ADDR2: 50, ADDR3: 50], or [ADDR1: 50%, ADDR2: 25%, ADDR3: 25%
    suite.assert_amount(1_000);
    suite.assert_ends_at(Expiration::AtHeight(1_000_000));
    suite.assert_duration(10);

    // skip 1/10th of the time
    suite.skip_blocks(100_000);

    // ADDR1 stakes additional 100 tokens, bringing the new staking power split to
    // [ADDR1: 200, ADDR2: 50, ADDR3: 50], or [ADDR1: 66.6%, ADDR2: 16.6%, ADDR3: 16.6%]
    // this means that per 100_000 blocks, ADDR1 should receive 6_666_666, while
    // ADDR2 and ADDR3 should receive 1_666_666 each.
    suite.mint_native_coin(coin(100, DENOM), ADDR1);
    suite.stake_native_tokens(ADDR1, 100);

    // rewards here should not be affected by the new stake,
    suite.assert_pending_rewards(ADDR1, DENOM, 5_000_000);
    suite.assert_pending_rewards(ADDR2, DENOM, 2_500_000);
    suite.assert_pending_rewards(ADDR3, DENOM, 2_500_000);

    // skip 1/10th of the time
    suite.skip_blocks(100_000);

    // here we should see the new stake affecting the rewards split.
    suite.assert_pending_rewards(ADDR1, DENOM, 5_000_000 + 6_666_666);
    suite.assert_pending_rewards(ADDR2, DENOM, 2_500_000 + 1_666_666);
    suite.assert_pending_rewards(ADDR3, DENOM, 2_500_000 + 1_666_666);
  
   <...>
}
```

Reply:
ok yes, sick.

memorializing what we talked about in discord: seems like this is working because `SnapshotMap`, which we use in staking modules to store historical voting power values, returns the *previous* block's value when queried at a given height, meaning that updates in hooks still access the previous voting power and total power values, even though they were just updated.

this is what is making this logic work, since really the rewards update needs to use the last voting power to update its pending rewards just before the new voting power takes effect.

however, we shouldn't rely on the staking contract (caller) to always give us the voting power/total power before the change in a hook, and instead we should be querying total power from the *previous* block, not the current one.